### ISSUE TITLE
AK47 Improvement: (Min) width between bullets 

### ISSUE THREAD

**Issue by Camille92** [ctubio/tribeca#10]
Hey Carles,

Just another idea to make Tribeca more profitable.

When using AK47, you're exposing more of your capital to the market, so I think it should be rewarded by more potential profits.

I don't know exactly how the position of bullets are calculated but sometime they stack together or are only 1c appart. (see picture).

It happened to me that when a big order hits the market movements (especially sell orders), all my bullets are eaten in one go at nearly the same price while the order went very deep in the book.

The idea would be to have the opportunity to fix a minimum width between each bullets to get the best of big orders.

It is especially important I think, now that I see many 0.01 orders "spoofing" the market and make tribeca blind.

One example to be more clear:
Width setting 2, width between bullets setting 0.2, 5 bullets

Bid price is €900 ask price is €902 so you're on top of the market both side and the bullets would act as usual but would be spaced by a minimum of €0.2 :
Bid: 899.8, 899.6, 899.4, 899.2, 899
Ask: 902.2, 902.4, 902.6, 902.8, 903

If in the order book at the bid side there is no quote for instance between 899.5 and 898 Tribeca would adjust the bullets in that manner:
900, 899.8, 899.6, 898, 897.8.

Tell me what you think and if it's easily doable,

![capture d ecran 2017-01-18 a 17 22 13](https://cloud.githubusercontent.com/assets/6861590/22072585/19b5667a-dda3-11e6-936b-4dc48456cd12.png)

Happy trading,
Camille

**Comment by ctubio** [comment]
Doable in theory! lets try. many many thanks for the idea Camille'''''

**Comment by ctubio** [comment]
![300px-spock_vulcan-salute3](https://cloud.githubusercontent.com/assets/1634027/22077151/4110e73e-ddb3-11e6-9d84-358e9f133d34.png)

**Comment by Camille92** [comment]
I'm happy you like it 😄  !

**Comment by ctubio** [comment]
needs to be optimized, but here it begins, v1 of `range`

**Comment by Camille92** [comment]
Hey I've seen it.

Thank you very much for the quick implementation.

I'll come back to you if I have an idea for optimization (and don't hesitate to ask about the difficulties you're encountering).

Have a nice day,

Camille

**Comment by ctubio** [comment]
Fixed wrong math for ask side at fc7725d

**Comment by Camille92** [comment]
(Re) Hello Carles,

I don't know if you have the same issue but the bullets don't act as they should in tribeca.

Here is a list of the issues (see pictures, 4 bullets using 0.2 range, no APR):
_ Sometime tribeca creates more bullets than set (pic 2).
_Sometime the bullets can "stack" together (on the same price) and at 1c appart (usually under APR but not always) (pic 1 and 2).
_ One or two bullets would go wayyyy down on the market over the fixed range. (pic 1)
_ Bullets would not fill the gap to go 1c above the next order on the book. (pic 3).

I don't like to raise issues without proposing an idea on how to fix the problem, so here is my thinking. From my understanding (correct me if i'm wrong) the bullet are now fixed by the previous bullet (or the first quote).

**Instead of that each bullet could act as one independent boomerang quote.**

In other word the position of a bullet would be define by:
Bid: FV - (width /2) - range*(rank of the bullet -1)
Ask: FV + (width/2) + range*(rank of the bullet -1)

Ie: Setting: width = 2, range = 0.5, and fv = 900:
The third bullet would be at (minimum):
900 - (2/2) - 0.5*(3 -1) = 888

Of course like for the "normal quote" if there is a gap on the market, the bullet should go just 1c over the next order like it is done in boomerang. 😃 

Img 1:
![capture d ecran 2017-01-24 a 11 58 08](https://cloud.githubusercontent.com/assets/6861590/22245155/d190f4aa-e22e-11e6-929b-40472f0bdd9f.png)

Img 2:
![capture d ecran 2017-01-24 a 14 41 06](https://cloud.githubusercontent.com/assets/6861590/22249462/55bd43a0-e243-11e6-9f13-6d2c1249e63f.png)

img 3: Logically the trade should be at 846.09 no ?
![capture d ecran 2017-01-24 a 17 54 35](https://cloud.githubusercontent.com/assets/6861590/22257436/5b0533d4-e25e-11e6-9cd0-212ef3442fa5.png)

I hope it all makes sense to you!

Have a nice day,

Camille

Updated: 2nd and 3rd picture

**Comment by Camille92** [comment]
![capture d ecran 2017-01-25 a 11 00 14](https://cloud.githubusercontent.com/assets/6861590/22285840/8c775e08-e2ed-11e6-9d17-bf14e86cd243.png)

I don't get it, sometimes it works perfectly and then it will try to put order 60€ under market price ahah.

Ps: You should post your btc address somewhere you deserve some tips for your work 😉

**Comment by ctubio** [comment]
yea i still dont understand the situation where it fails to de/increment by `range`, still under debug.

maybe because orders are placed too fast, and secondary orders do not even know the price of the first, but this seems unlikely because tribeca stores the price before send to exchange, and because orders are always send in serial, never in parallel... anyway still under debug'¡

**Comment by Camille92** [comment]
Okay, I'll try to look at that to see if i can figure something out of it.


Don't hesitate to tell me if I can do something to help.

**Comment by Camille92** [comment]
I found what seems to be the bug (or at least some of it), particularly about getting very low on the market.

If tribeca cancel (of if you do it manually by cancelling an order):

The order will reappear a range below the lowest order (didn't try for ask side) instead of staying at the same place.
Let me put you 2 pictures for you to see (here I clicked twice so the orders went €1 down:

![capture d ecran 2017-01-25 a 18 23 48](https://cloud.githubusercontent.com/assets/6861590/22301627/f8592528-e32b-11e6-8901-194bb4549443.png)
![capture d ecran 2017-01-25 a 18 24 04](https://cloud.githubusercontent.com/assets/6861590/22301630/fb0e9e60-e32b-11e6-970b-0df9ebeeaac1.png)

As well, if you cancel the last order, it will go a range below itself!

I hope it helps with debugging!
